lib/mongoidable/document_extensions.rb

```diff
@@ -6,10 +6,71 @@ module DocumentExtensions
     extend ActiveSupport::Concern
 
     included do
-      ability_class = Mongoidable.configuration.ability_class.constantize
+      ability_class = Mongoidable.configuration.ability_class
       raise TypeError unless ability_class == Mongoidable::Ability || ability_class.superclass == Mongoidable::Ability
 
-      embeds_many :instance_abilities, class_name: Mongoidable.configuration.ability_class
+      embeds_many :instance_abilities, class_name: Mongoidable.configuration.ability_class.to_s do
+        def update(**attributes)
+          return unless should_update?(attributes)
+
+          attributes = ActionController::Parameters.new(attributes) if attributes.is_a?(Hash)
+          query = ability_query_attrs(attributes)
+
+          #
+          ability = instance_abilities.where(query).first
+          type = ability_type(attributes[:action])
+          method = parent_document.new_record? ? :build : :create
+
+
+          create_attributes =  ability_create_attrs(attributes)
+          ability = instance_abilities.where(query).first || instance_abilities.send(method, create_attributes, type)
+
+          update_attrs = ability_update_attrs(ability, attributes)
+          ability.update_attributes update_attrs
+
+          ability
+        end
+
+        private
+
+        def should_update?(attributes)
+          inverse_method = attributes[:base_behavior] ? :cannot? : cannot?
+
+          parent_document.current_ability.send(inverse_method, *attributes.values_at(:action, :subject, :extra))
+        end
+
+        def ability_update_attrs(ability, query_params)
+          query_params.permit(ability.class.permitted_params).tap do |params|
+            params["extra"] =  query_params.require(:extra).permit!
+          end.to_h.with_indifferent_access
+        end
+
+        def ability_create_attrs(query_params)
+          query_params.permit(:action).tap do |params|
+            params["subject"] = Mongoidable::ClassType.mongoize(query_params["subject"])
+          end.merge(query_params.to_unsafe_hash.slice(:extra)).to_h.with_indifferent_access
+        end
+
+        def ability_query_attrs(query_params)
+          subject = Mongoidable::ClassType.mongoize(query_params[:subject])
+          query = query_params.permit(:action, :extra => []).merge(query_params.to_unsafe_hash.slice(:extra))
+          query["subject.type"] = subject[:subject]
+          query["subject.value"] = subject[:subject]
+          query
+        end
+
+        def ability_update_attrs(ability, query_params)
```

Comment:
Is this defined twice?  once here and once on line 42?

Maybe I'm missing/mis-reading something.